subsys/net/ip/tcp.c

```diff
@@ -3252,72 +3294,65 @@ int net_tcp_queue_data(struct net_context *context, struct net_pkt *pkt)
 		goto out;
 	}
 
-	len = net_pkt_get_len(pkt);
+	if (msg) {
+		len = 0;
 
-	if (conn->send_data->buffer) {
-		orig_buf = net_buf_frag_last(conn->send_data->buffer);
+		for (int i = 0; i < msg->msg_iovlen; i++) {
+			len += msg->msg_iov[i].iov_len;
+		}
 	}
 
-	net_pkt_append_buffer(conn->send_data, pkt->buffer);
-	conn->send_data_total += len;
-	NET_DBG("conn: %p Queued %zu bytes (total %zu)", conn, len,
-		conn->send_data_total);
-	pkt->buffer = NULL;
+	/* Queue no more than TX window permits. */
+	len = MIN(conn->send_win - conn->send_data_total, len);
 
-	ret = tcp_send_queued_data(conn);
-	if (ret < 0 && ret != -ENOBUFS) {
-		tcp_conn_close(conn, ret);
-		goto out;
-	}
-
-	if ((ret == -ENOBUFS) &&
-		(conn->send_data_total < (conn->unacked_len + len))) {
-		/* Some of the data has been sent, we cannot remove the
-		 * whole chunk, the remainder portion is already
-		 * in the send_data and will be transmitted upon a
-		 * received ack or the next send call
-		 *
-		 * Set the return code back to 0 to pretend we just
-		 * transmitted the chunk
-		 */
-		ret = 0;
-	}
+	if (msg) {
+		for (int i = 0; i < msg->msg_iovlen; i++) {
+			int iovlen = MIN(msg->msg_iov[i].iov_len, len);
 
-	if (ret == -ENOBUFS) {
-		/* Restore the original data so that we do not resend the pkt
-		 * data multiple times.
-		 */
-		conn->send_data_total -= len;
+			ret = tcp_pkt_append(conn->send_data,
+					     msg->msg_iov[i].iov_base,
+					     iovlen);
+			if (ret < 0) {
```

Comment:
If we could only queue partial data from msg, what should happen to the rest of it?

In previous version, we bailed out if we could not create net_pkt and returned error to the user. This is the way that is described in `sendmsg()` manual page.
```
If the message is too long to pass atomically through the underlying protocol,
the error EMSGSIZE is returned, and the message is not transmitted. 
```
But now we queue the partial data so this sending is no longer atomic.

Reply:
tl;dr; I don't think we modify the already existing behavior after all, see explaination below.

> In previous version, we bailed out if we could not create net_pkt and returned error to the user.

That's only correct for datagram sockets. For stream, it'd still be possible to send partial data, however we'd ignore the window size and use MTU as the delimeter of the maximum transfer length.
* The packet is allocated here (up to the MTU size):
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L1938
* Verify the allocated length here (but only fail in case of DGRAM sock):
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L1944
* And finally copy data to net_pkt, but only up to allocated space:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L2011

So, the behaviour was pretty much the same, the only difference is that now we respect the window size.

> >If the message is too long to pass atomically through the underlying protocol,
> >the error EMSGSIZE is returned, and the message is not transmitted. 
>
> But now we queue the partial data so this sending is no longer atomic.

That's one of those ambiguities of the POSIX documentation... Does that comment really apply to STREAM sockets? The same comment can be found in regular `send()` documentation. Plus, the error code description says EMSGSIZE only applies to socket types that require message atomicity (I'd interprate that as DGRAM sockets?):
> EMSGSIZE
The socket type requires that message be sent atomically, and the size of the message to be sent made this impossible.


> If we could only queue partial data from msg, what should happen to the rest of it?

The application would need to verify length sent, modify `msghdr` content, and try again. We do it like this for example in mqtt already (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/mqtt/mqtt_transport_socket_tcp.c#L108)

Reply:
For datagram sockets the situation is clear. But for stream socket the situation is more hazy as you say. I think earlier we made sure that the net_pkt was created properly (so that we had enough net_bufs) before returning to the user. It is totally different thing how the TCP queueing is done, from application point of view the message is now sent (even if it is queued).
Earlier in `context_write_data()` we tried to create full net_pkt and if it failed, then we returned error to the user without trying to send partial data.

Reply:
The documentation is indeed a bit vague, https://pubs.opengroup.org/onlinepubs/009604599/functions/sendmsg.html
It says for example that
```
Successful completion of a call to sendmsg() does not guarantee delivery of the message.
A return value of -1 indicates only locally-detected errors.

If space is not available at the sending socket to hold the message to be transmitted and
the socket file descriptor does not have O_NONBLOCK set, the sendmsg() function shall
block until space is available. If space is not available at the sending socket to hold the
message to be transmitted and the socket file descriptor does have O_NONBLOCK set,
the sendmsg() function shall fail.
```
So we could block if trying to send and there is no net_bufs available.
I checked also the code in zephyr tree and it seems that for example mqtt and websocket is trying to send partial data. But for example TLS socket did not.

Reply:
> So we could block if trying to send and there is no net_bufs available.

But what about the non-blocking case. For example, part of the message was already queued for sending, what then? We can't return EAGAIN, as part of the message was already queued, and we shoudn't really block either.

I'm not really a fan of returning error if the message size > TCP TX window size, because it inflicts artificial limits on stream sockets using sendmsg and creates ambiguity. What should the application do in case of such an error? It'd need to guess, whether it's ok to try again later, or maybe the message shall be trimmed as it exceeds the maximum TX window length (or formerly interface MTU). IMO, partial send is much less ambiguous in such case.

Plus it seems, that other implementations (Linux in this case) allow for partial `sendmsg` sends:
https://stackoverflow.com/questions/57792687/how-to-deal-with-sendmsg2-when-it-send-a-message-partially-in-linux

> But for example TLS socket did not.

That indeed looks a bit fishy TBH, could use some improvement. Seems potentially faulty in case of non-blocking socsk.

Reply:
Ok, I am convinced, partial sendmsg seems to be a thing although the documentation is a bit ambiguous. Lets go with this as it improves TCP a lot.

Reply:
Thanks!